### PR TITLE
LHA archive decompression support added

### PR DESCRIPTION
**Pull request by TwanVanDongen**
LHA Archive support added for decompression of methods lh0, lh5, lh6 and lh7. Others will be added when test-archives are created. Implementation taken from [LHA Decompressor for .NET](https://github.com/benbaker76/LHADecompressor).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,6 @@ tools
 .DS_Store
 *.snupkg
 /tests/TestArchives/6d23a38c-f064-4ef1-ad89-b942396f53b9/Scratch
+/src/SharpCompress/Readers/ReaderFactory.cs
```
Comment (adamhathcock) [inline .gitignore]: This doesn't seem right

File: src/SharpCompress/Archives/Lha/LhaArchive.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
+#nullable disable
```
Comment (adamhathcock) [inline src/SharpCompress/Archives/Lha/LhaArchive.cs]: Not a fan but can overlook for this PR

File: tests/SharpCompress.Test/SharpCompress.Test.csproj [hunk 3]
```diff
@@ -4,6 +4,9 @@
     <AssemblyName>SharpCompress.Test</AssemblyName>
     <PackageId>SharpCompress.Test</PackageId>
   </PropertyGroup>
+  <ItemGroup>
+    <Compile Remove="Lha\LhaReaderTests.cs" />
```
Comment (adamhathcock) [inline tests/SharpCompress.Test/SharpCompress.Test.csproj]: Seems wrong

**Review by adamhathcock — CHANGES_REQUESTED** [review]
Never heard of LHA!

Some minor things but the major is the LICENSE which is LGPL and we're MIT

Unless you got permission, I can't accept code from https://github.com/benbaker76/LHADecompressor/tree/master unfortunately

File: .gitignore (hunk 1 above)
Comment (TwanVanDongen) [reply .gitignore]: yep, accidentally removed it, but it's back now!

File: src/SharpCompress/Archives/Lha/LhaArchive.cs (hunk 2 above)
Comment (TwanVanDongen) [reply src/SharpCompress/Archives/Lha/LhaArchive.cs]: also not too happy about it, but saw that it was accepted elsewhere; certainly want to fix it where code is accepted

File: tests/SharpCompress.Test/SharpCompress.Test.csproj (hunk 3 above)
Comment (TwanVanDongen) [reply tests/SharpCompress.Test/SharpCompress.Test.csproj]: Started implementing also the reader function, but removed it in order to get the 'Archive' support in first.

**Comment by TwanVanDongen** [comment]
> Never heard of LHA!
> 
> Some minor things but the major is the LICENSE which is LGPL and we're MIT
> 
> Unless you got permission, I can't accept code from https://github.com/benbaker76/LHADecompressor/tree/master unfortunately

Received confirmation from Headkaze's to use his C# port of LHADecompressor (" I have no problem with you using my port. I believe it's the same license as the original source. If it complies then it shouldn't be a problem."). This however only resolved the porting, now trying to connect the original Java developer (Nobuyasu SUEHIRO)  with the same question.

**Comment by vpenades** [comment]
Suggestion:  If the problem is only the license, an alternative approach would be this:

Create a separate project library called **SharpCompress.ArchiveFormat.LHA** (or something like that) which would use the LGPL license, so it would be license compatible.

Users willing to use LHA would require:

- To be also license compatible with LGPL
- Reference SharpCompress
- Reference SharpCompress.ArchiveFormat.LHA
- register the LHA factory at application startup

**Comment by adamhathcock** [comment]
that's a good call, to do that.  Can register factories at start up now with little code

**Pull request closed without merging**